core/src/consensus/tendermint/mod.rs

```diff
@@ -141,6 +140,34 @@ impl Tendermint {
         self.height.load(AtomicOrdering::SeqCst)
     }
 
+    pub fn update_future_proposal(&self, future_proposal: Bytes) {
+        let prev_score = self
+            .future_proposal
+            .read()
+            .as_ref()
+            .map(|prev| BlockView::new(prev).header_view().score())
+            .unwrap_or_default();
+        let new_score = BlockView::new(&future_proposal).header_view().score();
+        if prev_score < new_score {
+            ctrace!(ENGINE, "future proposal changed");
+            *self.future_proposal.write() = Some(future_proposal);
+        } else {
+            ctrace!(ENGINE, "future proposal not changed");
+        }
+    }
+
+    pub fn future_proposal_prev_hash(&self) -> Option<H256> {
+        if let Some(future_proposal) = self.future_proposal.try_read_for(std::time::Duration::new(1, 0)) {
```

Comment:
Why do you use `try_read_for`?
It should be used only when there is a fallback when it failed to acquire a lock.

Reply:
I'll fix it.